### PR TITLE
thumb : limit flickering in culling & preview

### LINKED ISSUE #7685: lighttable flickers with mini-image in preview mode
**Describe the bug**
Moving through images in preview mode results in a very rapid flicker of a tiny  crop of the image, then the full image shows up.

**To Reproduce**
1. Start darktable
2. Import a folder of images
3. Be sure that are in lighttable view in file manager mode
4. Select the first image
5. Press and hold W (or press shift-W)
6. Press right arrow to move through the images
7. A crop of the upcoming image appears, then the full image appears a moment later.

**Expected behavior** (best case)
- If the image is available in the cache at the needed size, it is displayed.
- If the cached image needs to be generated, the highest lower resolution mipmap available will be displayed, scaled to the needed size. The best resolution mipmap will be displayed once it is generated.

**Screenshots**
Here is a screenshot of a "mini" image of a solid green 6000x6000px rectangle (mini image is 46x46 pixels):

image

Here is a "mini" image of a raw file of a 4416x2944 photograph (mini-image is 74x46 pixels):

image

**Platform (please complete the following information):**
 - Darktable Version: 56a44ff605b86372493cbe738f9536d0ffec1c19 (master)
 - OS: Debian Linux
 - OpenCL: yes
 - Nvidia Quadra P4000, display driver v455.45.01

**Additional context**
- I've tested this with raw files and 6000x6000 pixel PNGs
- The result seems to be approximately the aspect ratio of the source image, but is cropped
- Can reproduce with a clean install
- The problem can go away once the cache is "warm"
- Flicker does not occur when moving between images in lighttable at max zoom
- Regardless of this bug, I've wondered for a while if, when a fallback lower-res image is displayed in lighttable, whether it should be scaled up to the size of the eventual adequate-res mipmap. I believe that currently it is displayed 1:1 which creates flicker.

### PR DESCRIPTION
**Pull request by AlicVB**
this fix #7685 

Not so many change, but 2 days of work and lot of headaches ;)

To sum up for those interested : creating thumb with the right size wasn't too hard, but then we face another problem, where the allocation widget size differ from the widget size at first draw (not after) this seems due to the fact that gtk seems to throw draw event during the widget resizing, in particular for those which autosize to their parent... => So we need to set the exact right size *before* showing them... right !

Then comes a margin problem for the image which seems to change between creation time and "shown" time... In fact that was due to the css which define smaller margins for images if they are inside a "fullpreview" container. But of course they aren't at creation time, but when they are packed into the container hence the margin change... (seems obvious, but take me 4 hours to find !)

### CODE REVIEW

**Comment by dtorop** [comment]
Mark me as very impressed by this work.

I see one glitch still: If I start up darktable and enter preview mode, I still see a mini-thumbnail flicker before the widget goes fullscreen. After that everything is great -- no flicker.

**Comment by AlicVB** [comment]
> If I start up darktable and enter preview mode, I still see a mini-thumbnail flicker before the widget goes fullscreen

yes, I know, but I don't know how to avoid it for now... This is due to the fact that when we create the thumb, we have no idea of their definitive size, except if there were previous thumbs...
That said, it's not as problematic as your issue imho...

**Comment by dtorop** [comment]
>  That said, it's not as problematic as your issue imho...

That's definitely true. A little flicker on entering preview mode is minor. The flicker when paging through preview images is distracting when trying to pick/rate them.

**Comment by dtorop** [comment]
One thought: Is it possible to determine the initial thumbnail dimensions, in the case of mode `DT_CULLING_MODE_PREVIEW`, by examining the dimensions of the allocation of culling's GtkLayout widget, or even the center widget?

**Comment by dtorop** [comment]
Here is a proof of concept for this:

```c
modified   src/dtgtk/culling.c
@@ -1147,6 +1147,11 @@ static gboolean _thumbs_recreate_list_at(dt_culling_t *table, const int offset)
         nw = th_model->width;
         nh = th_model->height;
       }
+      else if(table->mode == DT_CULLING_MODE_PREVIEW)
+      {
+        nw = table->view_width;
+        nh = table->view_height;
+      }
       dt_thumbnail_t *thumb;
       if(table->mode == DT_CULLING_MODE_PREVIEW)
         thumb = dt_thumbnail_new(nw, nh, nid, nrow, table->overlays, DT_THUMBNAIL_CONTAINER_PREVIEW,
@@ -1218,6 +1223,11 @@ static gboolean _thumbs_recreate_list_at(dt_culling_t *table, const int offset)
             nw = th_model->width;
             nh = th_model->height;
           }
+          else if(table->mode == DT_CULLING_MODE_PREVIEW)
+          {
+            nw = table->view_width;
+            nh = table->view_height;
+          }
           dt_thumbnail_t *thumb;
           if(table->mode == DT_CULLING_MODE_PREVIEW)
             thumb = dt_thumbnail_new(nw, nh, nid, nrow, table->overlays, DT_THUMBNAIL_CONTAINER_PREVIEW,
```

Here this works to get rid of the last flicker , the one when entering preview mode.

It looks as though `dt_culling_full_redraw()` calls `_compute_sizes()`, which initializes view_width/view_height before it calls `_thumbs_recreate_list_at()` so it should be possible to rely on them for the dimensions. It could be possible to even rely on that value entirely? Example:

```diff
modified   src/dtgtk/culling.c
@@ -1140,7 +1140,12 @@ static gboolean _thumbs_recreate_list_at(dt_culling_t *table, const int offset)
       // draw events
       int nw = 40;
       int nh = 40;
-      if(g_list_length(table->list) > 0)
+      if(table->mode == DT_CULLING_MODE_PREVIEW)
+      {
+        nw = table->view_width;
+        nh = table->view_height;
+      }
+      else if(g_list_length(table->list) > 0)
       {
         dt_thumbnail_t *th_model
             = (dt_thumbnail_t *)g_list_nth_data(table->list, MIN(pos, g_list_length(table->list) - 1));
@@ -1211,7 +1216,12 @@ static gboolean _thumbs_recreate_list_at(dt_culling_t *table, const int offset)
           // trigger draw events
           int nw = 40;
           int nh = 40;
-          if(g_list_length(table->list) > 0)
+          if(table->mode == DT_CULLING_MODE_PREVIEW)
+          {
+            nw = table->view_width;
+            nh = table->view_height;
+          }
+          else if(g_list_length(table->list) > 0)
           {
             dt_thumbnail_t *th_model
                 = (dt_thumbnail_t *)g_list_nth_data(table->list, MIN(pos, g_list_length(table->list) - 1));
```

Is there ever a case in preview mode when the code falls back to the "in rare cases" code to create another thumbnail? I don't know the code well enough to tell -- or indeed much of this!

**Comment by AlicVB** [comment]
@dtorop : great thanks !
I've added that in my last commit. I've referred to you in the msg, but if you prefer to do it on your own, no problem !

**Comment by dtorop** [comment]
@AlicVB Glad that you added the commit. Thanks for asking about who should do it -- I'm quite happy you did.

Not for this PR, I'm still curious if this code path:

https://github.com/darktable-org/darktable/blob/04eca63b3a94fec764921851122fc2015b88574e/src/dtgtk/culling.c#L1156-L1160

ever gets used in preview mode. If not, it could be simplified?

**Comment by AlicVB** [comment]
@dtorop : you're right, that's not possible.
I've removed the check +  added a new way to retrieve size from other thumbs (this cover the case where we have thumb with right size in the new list but the old list is empty, for example if all their thumb have already been put inside the new list)

**Comment by xfh** [comment]
I still notice an occasional flickering on your latest commit: 
https://user-images.githubusercontent.com/9366771/103954843-58d6d680-5145-11eb-8c57-bf5b2ad8302d.mp4

I am not sure why it happens. I've noticed in the debugger, that the initial image_w widget is a bit too large, because it does not account for margins. Later on image_w and w_image_box shrink a bit in https://github.com/darktable-org/darktable/blob/9dba538639cb7345b62b495e9c1d7b0abf146055/src/dtgtk/thumbnail.c#L349-L358

Maybe I am sometimes switching images so fast, that a 1st preview is rendered but the final, correctly scaled thumbnail is skipped?
Here is another recording with 100% zoom:
https://user-images.githubusercontent.com/9366771/103955549-ce8f7200-5146-11eb-922c-d93b4a2f01d8.mp4

**Comment by AlicVB** [comment]
@xfh : I reproduce. Interestingly this seems to only happen with vertical images and the shift is only on horizontal "margins", not vertical, no size change... I investigate !

**Comment by AlicVB** [comment]
I've finally found the cause and why I didn't cache the problem at first : we rely on saved aspect_ratio to determine size before we load the image itself, and aspect_ratio... is round to 1 decimal when saved in the database :( (this is needed for collect)
So for "big" sizes (the case in preview) we get noticeable errors...

Now I need to find a way to solve that without adding too much computation... Definitely I'll be happy when this issue is solve !

**Comment by xfh** [comment]
Hmm, you don't think it has to do with the margins? You might want to have a look at my latest changes in #7696. It looks to me like the image is initially a bit too large (due to no margins) and then scaled down (once margins are added).

**Comment by AlicVB** [comment]
no, this "remargin" is not a real problem, as at the end of dt_thumbnail_create_widget we resize all the widgets properly with right margins and so... before to show anything.
But the resizing is based on the aspect ratio set at first.

As an example, with "classical" 3:2 images, you won't see any flickering with landscape format (ratio = 1.5) but do see it with portrait image (ratio = 0.66666)

**Comment by AlicVB** [comment]
@xfh : can you test with last commit, it should fix your flickering with vertical images (or any images which needs precise ratio)
Note that I haven't done anything for zoomed images, as you seems to have already done it in your PR...

For the record, I recompute the ratio from the largest available mipmap if any. This doesn't seems to have any noticeable speed effect (timings stay below 0.001 s here)

File: src/dtgtk/thumbnail.c [hunk 1]
```diff
@@ -1098,26 +1098,30 @@ GtkWidget *dt_thumbnail_create_widget(dt_thumbnail_t *thumb)
...
+            iw = ih * ar;
+          else
+            ih = iw / ar;
+          break;
+        }
       }
-      // rescale to ensure it stay in thumbnails bounds
-      const float scale = fminf(1.0, fminf((float)thumb->width / iw, (float)thumb->height / ih));
-      iw *= scale;
-      ih *= scale;
```
Comment (xfh) [inline src/dtgtk/thumbnail.c]: Why did you remove the scaling?

File: src/dtgtk/thumbnail.c [hunk 2]
```diff
@@ -1098,26 +1098,30 @@ GtkWidget *dt_thumbnail_create_widget(dt_thumbnail_t *thumb)
...
     int iw = thumb->width;
     int ih = thumb->height;
-    if(img)
+    // we can't rely on img->aspect_ratio as the value is round to 1 decimal, so not enough accurate
+    // so we compute it from the larger available mipmap
+    for(int k = DT_MIPMAP_7; k >= DT_MIPMAP_0; k--)
     {
-      const float ar = img->aspect_ratio;
-      dt_image_cache_read_release(darktable.image_cache, img);
-      if(ar > 0.0001)
```
Comment (xfh) [inline src/dtgtk/thumbnail.c:1102]: Strange, my `ar` did always had more than 1 decimal, even before your change.

**Review by xfh — COMMENTED** [review]
@AlicVB I've just tested your latest changes, but I still have the issue. Also with landscape images (I am switching between duplicates with different white balance).

Correct position (image1):
cap_landscape-flickering3_00:00:02_01

This image causes the flickering:
cap_landscape-flickering3_00:00:02_02

Finally, the correct position is restored:
cap_landscape-flickering3_00:00:02_03

It's hard to spot since it's a subtle shift, but it's there. Maybe more obvious in a video:
https://user-images.githubusercontent.com/9366771/104070276-ca2a8e00-5206-11eb-911c-d0297f2761ce.mp4

File: src/dtgtk/thumbnail.c (hunk 2 above)
Comment (xfh) [reply src/dtgtk/thumbnail.c:1102]: Okay, that seems to depend on the image. The portrait I've used before had aspect ratio from image cache: 0.74545455, with your change it's 0.75 *and* I don't notice any flickering. Seems like you are on the right track, the correct aspect ratio is indeed 0.75 :tada: 

The portrait above has aspect ratio from image cache: 1.5
With your change, it is 1.50916672.
Once I've zoomed to 100% and re-open the image, the final_width and final_height get initialised. Then the aspect ratio is
When I take `(float)img->final_width / img->final_height` it is 1.50884688.

Conslusion: the mipmap cache improves the situation, but it is not always accurate enough.

Comment (xfh) [reply src/dtgtk/thumbnail.c:1102]: I figured out why it still had a flicker. It's a rounding problem!

I'll try to explain based on my sample image:
| value      | width | height |
| ----------- | ------- | ------ |
| thumb    | 1920 | 1139 |
| mipmap | 1811 | 1200 |
| initial thumb->w_image (in dt_thumbnail_create_widget) | 1920 | 1272 | 
| new thumb->w_image (in _thumb_set_image_area) |  1686.03784 | 1117 |
| cairo_image_surface_get_width(thumb->img_surf) (in _event_image_draw) | 1685 | 1117 |

The problem is in _thumb_set_image_area, where the width is scaled down when taking margins into account. I have in this example `image_w = 1882` and `image_h = 1117` and the initial w_image sizes. The resulting scale is
 ```const float scale = fminf((float)image_w / wi, (float)image_h / hi) = 0.878144681```.
The new image_w width is 1 px too large compared to what the final cairo calculated width returns.

Interestingly, when I revert your removal of the scaling (see https://github.com/darktable-org/darktable/pull/7719#discussion_r554220593), then I don't have any more flickering :tada: 

After adding the scaling, the initial width/height of thumb->w_image is 1719 x 1139.
That gives later in _thumb_set_image_area `scale = 0.980684817` and new thumb->w_image size 1685.79724 x 1117.

I suggest to re-add the scaling. The cairo surface is scaled in a similar way:
https://github.com/darktable-org/darktable/blob/master/src/views/view.c#L963-L967

**Comment by AlicVB** [comment]
@xfh : thanks for the review and tests. I've made the initail size more robust, with the help of your comments : 
- take account of margins from the beginning (note that this margins are "minimal" ones, they can be greater with some overlays, but that's a good start at least)
- use float to compute the initial size and so avoid a little much rounding errors. That allow to put back the "fit" scaling too.

Now we have to keep in mind that there's still cases where this can fail : if you don't have cached image, you use very small thumbs in filemanager, you enter preview and immediately and quickly switch between images (to not let mipmap preloading occurring) : in that case, we only have very small mipmap to compute aspect_ratio and -again- this may not ends to rounding errors and small flickering. But well... this is quite a corner case, and this only solution I see to avoid this would be to get the "real" size to compute the ratio, and that would mean noticeable slowdown... I prefer flickering ;)

**Comment by AlicVB** [comment]
last commit refactor so even other overlays than "block" avoid flickering...

**Review by xfh — COMMENTED** [review]
@AlicVB I've tested with your latest change. As you already mentioned, there are still cases, where flickering can be observed due to the rounding problem, but I don't see how this can be solved and it's not really worth the trouble. Your changes are already a great improvement. Thanks!

One case I would like to mention is in filemanager, when only showing 1 thumb. It seems like the allocated space for the overlays changes. 

It seems independent of the type of overlay. Maybe something for another issue?

https://user-images.githubusercontent.com/9366771/104103624-673b0480-52a3-11eb-8530-b6e25aa6a2cf.mp4

https://user-images.githubusercontent.com/9366771/104103627-686c3180-52a3-11eb-9b4c-6928f3918ad0.mp4

**Comment by AlicVB** [comment]
@xfh : the issue with lighttable and zoom == 1 was the same problem as for preview at first : the css that was not right until the widget is put in his parent (but it's too late) My last commit should have fixed it

**Review by xfh — APPROVED** [review]
Great, I can confirm that this fixes the overlay resizing!

**Comment by TurboGit** [comment]
Thanks for the hard work on this one!

**Comment by chhil** [comment]
Not sure if its related, or something can be done for the flickering when I click the hamburger for presets for modules in darkroom and the selection menu pops up.

**Comment by AlicVB** [comment]
@chhil : afaict, this is completely unrelated... Please open a new issue and describe exactly what happens Thanks

**Pull request merged**